receiver/kafkareceiver/consumer_franz.go

```diff
@@ -0,0 +1,476 @@
+		)
+		go func(pc *pc, msgs []*kgo.Record) {
+			defer wg.Done()
+			defer pc.wg.Done()
+			fatalOffset := int64(-1)
+			var lastProcessed *kgo.Record
+			for _, msg := range msgs {
+				if !c.config.MessageMarking.After {
+					c.client.MarkCommitRecords(msg)
+				}
+				if err := c.handleMessage(pc, wrapFranzMsg(msg)); err != nil {
+					pc.logger.Error("unable to process message",
+						zap.Error(err),
+						zap.Int64("offset", msg.Offset),
+					)
+					// To keep both Sarama and Franz implementations consistent,
+					// we pause consumption for partitions that have fatal errors,
+					// which isn't ideal since there needs to be some sort of manual
+					// intervention to unlock the partition. But this is already
+					// mentioned in the docs and also happens with Sarama.
+					if !c.config.MessageMarking.OnError {
+						fatalOffset = msg.Offset
+						break // Stop processing messages.
+					}
+				}
+				lastProcessed = msg // Store so we can commit later.
+			}
+			// Pause topic/partition processing locally, any rebalances that move
+			// away the process the partition regularly, which will re-process
+			// the message.
+			if fatalOffset > -1 {
+				c.client.PauseFetchPartitions(map[string][]int32{
+					p.Topic: {p.Partition},
+				})
+				// We don't return false since we want to avoid shutting down
+				// the consumer loop and consumption due to message poisoning.
+				// If we did, we would cause an eventual systematic failure if
+				// there are more topic / partitions in this consumer group when
+				// the partition is rebalanced to another consumer in the group.
+				//
+				// Ideally, we would attempt to re-process permanent errors
+				// for up to N times and then pause processing, or even better,
+				// produce the message to a dead letter topic.
+				pc.logger.Error("unable to process message: pausing consumption of this topic / partition on this consumer instance due to MessageMarking.OnError=false",
+					zap.Int64("offset", fatalOffset),
+				)
+			}
+			if lastProcessed == nil {
+				return // No metrics nor marks to update.
+			}
+			// Otherwise, publish consumer lag.
+			c.telemetryBuilder.KafkaReceiverOffsetLag.Record(
+				context.Background(),
+				(p.HighWatermark-1)-(lastProcessed.Offset),
+				metric.WithAttributeSet(pc.attrs),
+			)
+			if c.config.MessageMarking.After {
+				c.client.MarkCommitRecords(lastProcessed)
+			}
+		}(assign, p.Records)
+	})
+	// Wait for all records to be processed and commit if autocommit=false.
+	wg.Wait()
+	if !c.config.AutoCommit.Enable {
+		if err := c.client.CommitMarkedOffsets(ctx); err != nil {
+			c.settings.Logger.Error("failed to commit offsets", zap.Error(err))
+		}
+	}
+	return true
+}
+
+func (c *franzConsumer) Shutdown(ctx context.Context) error {
+	if !c.triggerShutdown() {
+		return nil
+	}
+
+	select {
+	case <-ctx.Done():
+		c.cancelProcessing(fmt.Errorf(
+			"kafka consumer: shutdown context done: %w", ctx.Err(),
+		))
+		return ctx.Err()
```

Comment:
I just noticed that this is the only time we cancel `processingCtx`. I'm not convinced that `processingCtx` is needed. Would the following work?
 - remove `processingCtx` and `cancelProcessing`, just pass `context.Background()` into `consumeLoop`
 - in `consumeLoop`, rely on the client method calls being interrupted by closing the client in `triggerShutdown`
 - in `handleMessage`, rely on the partition context being cancelled by the lost hooks when closing the client in `triggerShutdown`